src/Adapter/Carrier/Repository/CarrierRangeRepository.php

```diff
@@ -0,0 +1,249 @@
+     *
+     * @return CarrierRange[]
+     *
+     * @throws \PrestaShop\PrestaShop\Core\Domain\Zone\Exception\ZoneException
+     */
+    public static function formatRangesFromData(IteratorAggregate|array $iterator): array
+    {
+        $ranges = $prices = [];
+        $tmpLast = ['from' => null, 'to' => null];
+        foreach ($iterator as $range) {
+            if ($tmpLast['from'] !== $range['range_from'] || $tmpLast['to'] !== $range['range_to']) {
+                if (count($prices) > 0) {
+                    $ranges[] = new CarrierRange(
+                        (float) $tmpLast['from'],
+                        (float) $tmpLast['to'],
+                        $prices
+                    );
+                    $prices = [];
+                }
+                $tmpLast = ['from' => $range['range_from'], 'to' => $range['range_to']];
+            }
+
+            $prices[] = new CarrierRangePrice(
+                (int) $range['id_zone'],
+                (float) $range['range_price']
+            );
+        }
+
+        $ranges[] = new CarrierRange(
+            (float) $tmpLast['from'],
+            (float) $tmpLast['to'],
+            $prices
+        );
+
+        return $ranges;
+    }
+
+    private function reset(CarrierId $carrierId, ShopConstraint $shopConstraint): void
+    {
+        $this->assertShopConstraint($shopConstraint);
+        $tablesToReset = [
+            'range_weight',
+            'range_price',
+            'delivery',
+        ];
+
+        foreach ($tablesToReset as $table) {
+            $this->connection->delete(
+                $this->dbPrefix . $table,
+                ['id_carrier' => $carrierId->getValue()]
+            );
+        }
+    }
+
+    private function getRangeTypeTable(Carrier $carrier): string
+    {
+        switch ($carrier->shipping_method) {
+            case Carrier::SHIPPING_METHOD_WEIGHT:
+                return 'range_weight';
+            case Carrier::SHIPPING_METHOD_PRICE:
+                return 'range_price';
+            default:
+                throw new CarrierException(sprintf('Unknown shipping method "%s"', $carrier->shipping_method));
+        }
+    }
+
+    private function applyRangeTypeForQuery(QueryBuilder $queryBuilder, Carrier $carrier): QueryBuilder
+    {
+        // Define which table to join based on carrier shipping method
+        $tableRange = $this->getRangeTypeTable($carrier);
+
+        // Join the range table and order by range
+        $queryBuilder->innerJoin(
+            'cd',
+            $this->dbPrefix . $tableRange,
+            'cr',
+            'cd.id_' . $tableRange . ' = cr.id_' . $tableRange . ' AND cr.id_carrier = :carrierId'
+        )
+            ->orderBy('cr.id_' . $tableRange, 'ASC');
+
+        return $queryBuilder;
+    }
+
+    private function applyShopConstraint(QueryBuilder $queryBuilder, ShopConstraint $shopConstraint): QueryBuilder
+    {
+        // todo: make shop constraint magic here!
```

Comment:
Ok for first implementation, but we'll have to be careful to make this compatible with already existing data

Especially when we have multiple shops, if the data keeps `id_shop` and `id_shop_group` null, it's easy because it's the "all shops" structure and it can be used as the reference for any shop

But what if data was duplicated for all the available shops with `id_shop = 1`, `id_shop = 2`, ... Then we'll need to make sure we only get one subset for one shop and make sure they are all equals, if not then what is the reference we're supposed to use?